src/Datadog.Trace.ClrProfiler.Managed/CallTarget/Handlers/IntegrationMapper.cs

```diff
@@ -0,0 +1,714 @@
+            }
+
+            // Load the exception
+            ilWriter.Emit(OpCodes.Ldarg_1);
+
+            // Load the state
+            ilWriter.Emit(OpCodes.Ldarg_2);
+
+            // Call Method
+            onMethodEndMethodInfo = onMethodEndMethodInfo.MakeGenericMethod(callGenericTypes.ToArray());
+            ilWriter.EmitCall(OpCodes.Call, onMethodEndMethodInfo, null);
+
+            ilWriter.Emit(OpCodes.Ret);
+
+            Log.Debug($"Created EndMethod Dynamic Method for '{integrationType.FullName}' integration. [Target={targetType.FullName}]");
+            return callMethod;
+        }
+
+        internal static DynamicMethod CreateEndMethodDelegate(Type integrationType, Type targetType, Type returnType)
+        {
+            Log.Debug($"Creating EndMethod Dynamic Method for '{integrationType.FullName}' integration. [Target={targetType.FullName}, ReturnType={returnType.FullName}]");
+            MethodInfo onMethodEndMethodInfo = integrationType.GetMethod(EndMethodName, BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static);
+            if (onMethodEndMethodInfo is null)
+            {
+                return null;
+            }
+
+            if (onMethodEndMethodInfo.ReturnType.GetGenericTypeDefinition() != typeof(CallTargetReturn<>))
+            {
+                throw new ArgumentException($"The return type of the method: {EndMethodName} in type: {integrationType.FullName} is not {nameof(CallTargetReturn)}");
+            }
+
+            Type[] genericArgumentsTypes = onMethodEndMethodInfo.GetGenericArguments();
+            if (genericArgumentsTypes.Length < 1 || genericArgumentsTypes.Length > 2)
+            {
+                throw new ArgumentException($"The method: {EndMethodName} in type: {integrationType.FullName} must have the generic type for the instance type.");
+            }
+
+            ParameterInfo[] onMethodEndParameters = onMethodEndMethodInfo.GetParameters();
+            if (onMethodEndParameters.Length < 3)
+            {
+                throw new ArgumentException($"The method: {EndMethodName} with {onMethodEndParameters.Length} paremeters in type: {integrationType.FullName} has less parameters than required.");
+            }
+            else if (onMethodEndParameters.Length > 4)
+            {
+                throw new ArgumentException($"The method: {EndMethodName} with {onMethodEndParameters.Length} paremeters in type: {integrationType.FullName} has more parameters than required.");
+            }
+
+            if (onMethodEndParameters[1].ParameterType != typeof(Exception) && onMethodEndParameters[2].ParameterType != typeof(Exception))
+            {
+                throw new ArgumentException($"The Exception type parameter of the method: {EndMethodName} in type: {integrationType.FullName} is missing.");
+            }
+
+            if (onMethodEndParameters[2].ParameterType != typeof(CallTargetState) && onMethodEndParameters[3].ParameterType != typeof(CallTargetState))
```

Comment:
Accessing `onMethodEndParameters[3]` may cause an out of bounds exception because the length can be 3 here

Reply:
fixed in https://github.com/DataDog/dd-trace-dotnet/pull/1007/commits/04ae61a340872f36f5e93929e69b26305bc991dd